iina/HardwareDecodeCapabilities.swift

```diff
@@ -0,0 +1,103 @@
+
+  private var initialization: DispatchWorkItem?
+  private var isInitialized = false
+
+  /// Cache containing a map from video codec to whether hardware decoding is supported.
+  private var supported: [CMVideoCodecType: Bool] = [:]
+
+  /// Check the hardware decoding capabilities of this Mac and cache the results.
+  ///
+  /// As checking the capabilities takes a long time it is performed asynchronously. See `isHardwareDecodeSupported` for details.
+  /// - Important: This method **must** be called before `isSupported` is called and only be called **once**.
+  func checkCapabilities() {
+    guard initialization == nil else {
+      // Internal error. This method must only be called once.
+      Logger.fatal("HardwareDecodeCapabilities is already initialized")
+    }
+    initialization = DispatchWorkItem() { [self] in
+      for codec in codecs {
+        supported[codec] = isHardwareDecodeSupported(codec)
+      }
+    }
+    DispatchQueue.global(qos: .userInitiated).async { self.initialization!.perform() }
+  }
+
+  /// Whether this Mac supports hardware decoding for the given video codec.
+  /// - Parameter codecType: The video codec as a
+  ///     [CMVideoCodecType](https://developer.apple.com/documentation/coremedia/cmvideocodectype).
+  /// - Returns: `true` if hardware decoding is supported;,`false` otherwise.
+  func isSupported(_ codecType: CMVideoCodecType) -> Bool {
+    if !isInitialized {
+      guard let initialization = initialization else {
+        // Internal error. The cache must be initialized before calling this method.
+        Logger.fatal("HardwareDecodeCapabilities.checkCapabilities has not been called")
+      }
+      initialization.wait()
+      isInitialized = true
+    }
+    guard let supported = supported[codecType] else {
+      // Internal error. This codec must not be in the list of codecs above.
+      Logger.fatal("HardwareDecodeCapabilities is missing support for codec \(codecType)")
+    }
+    return supported
+  }
+
+  /// Whether this Mac supports hardware decoding for the given video codec.
+  /// - Parameter codecType: The video codec as a
+  ///     [CMVideoCodecType](https://developer.apple.com/documentation/coremedia/cmvideocodectype).
+  /// - Returns: true` if hardware decoding is supported;,`false` otherwise.
+  /// - Important: This method calls
+  ///     [VTIsHardwareDecodeSupported](https://developer.apple.com/documentation/videotoolbox/vtishardwaredecodesupported(_:)),
+  ///     which if called on the main thread will cause Xcode to report: "This method should not be called on the main thread as it may
+  ///     lead to UI unresponsiveness". Use a different thread to call this method. Do not call `isHardwareDecodeSupported`
+  ///     from the main thread.
+  private func isHardwareDecodeSupported(_ codecType: CMVideoCodecType) -> Bool {
+    if #available(macOS 11.0, *) {
+      VTRegisterSupplementalVideoDecoderIfAvailable(codecType)
+    }
+    if #available(macOS 10.13, *) {
```

Comment:
Given that IINA now will only run on macOS 10.15+, we can safely remove this version check and the code after this, since they will never be executed.

Reply:
Fixed